elyra-ai/elyra
# Add support for ELYRA_RUN_NAME environment variable

This PR introduces a new protected environment variable `ELYRA_RUN_NAME`, which is made available to notebooks | scripts during pipeline execution. The value is a unique string that changes for every pipeline run. For example
- run 1: `ELYRA_RUN_NAME`=`some_value`
- run 2: `ELYRA_RUN_NAME`=`some_other_value`

### Local runtime environment

`ELYRA_RUN_NAME` is set to `<pipeline-name>-<timestamp>`

Example value for pipeline file `give_run_name.pipeline`:

```
give_run_name-0617153527
```


### KFP runtime environment

`ELYRA_RUN_NAME` is set to the KFP `<run_name>`

Example value for pipeline file `give_run_name.pipeline`:

```
give_run_name-0617153855
```

See screen captures in https://github.com/elyra-ai/kfp-notebook/pull/102

### Airflow runtime environment

`ELYRA_RUN_NAME` is set to `<pipeline-name-timestamp>-<AA-DAG-run-id>`

Example value for pipeline file `give_run_name.pipeline`:

```
give_run_name-0617150717-scheduled__2021-06-16T00:00:00+00:00
```
Example DAG:
![image](https://user-images.githubusercontent.com/13068832/122807318-43331480-d280-11eb-96f3-8adedb3aed86.png)

Task Info:

![image](https://user-images.githubusercontent.com/13068832/122807402-678ef100-d280-11eb-80b4-5b4a0a2a2ecf.png)

Job output (scheduled run):

![image](https://user-images.githubusercontent.com/13068832/122807535-90af8180-d280-11eb-9aa4-74c75992c0a5.png)

Job output (subsequent manual run)

![image](https://user-images.githubusercontent.com/13068832/122807900-00257100-d281-11eb-9dc6-5c9a400a8e65.png)



Fixes #1702

Notes:
- The actual value is the same for each node during pipeline execution.
- The actual value starts with the pipeline name, e.g. `my-pipeline-` followed by a sequence of characters that assure uniqueness.
- The trailing character sequence has either no semantic meaning (local execution) or is an run identifier (generated by KFP and AA). 
- The actual value _does not_ provide information about the order in which the runs where executed. 
- The actual value is not persisted by Elyra.
- If the user manually overrides the variable value in the node properties (e.g. sets `ELYRA_RUN_ID` to `bogusid`), the change is ignored, as it should be for all protected/system Elyra environment variables.

TODO:

- [x] Add documentation (https://github.com/elyra-ai/elyra/issues/1703)
- [x] Add support in local pipeline execution
- [x] Add support in KFP with Argo pipeline execution 
- [x] Add support in KFP with Tekton pipeline execution 
- [x] Add support in AA pipeline execution 

Dependencies:

- [x] [PR in `kfp-notebook` repo](https://github.com/elyra-ai/kfp-notebook/pull/102) (requires new published version of `kfp-notebook`)

### How was this pull request tested?

- [Download/extract the test case archive](https://github.com/elyra-ai/elyra/files/6658749/1702_testcase.tar.gz)
  into a JL workspace
- Open `1702_test.pipeline` using the VPE (note that the comments indicate which nodes attempt to override the environment variable value - n/a for local)
- Run the pipeline locally, on KFP, and AA and inspect the outputs.
- [KFP/AA] Export the pipelines, upload, and run them
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.
